### PR TITLE
feat(network) add nodeComponent to network

### PR DESCRIPTION
**Pull request by SheldonNunes**
The purpose of this PR is to introduce a new prop `nodeComponent` to @nivo/network. This `nodeComponent` allows for the overriding of the default Node component allowing for more customisation of networks. This has been a requested feature in a [previously stale issue](https://github.com/plouc/nivo/issues/833).

Also noticed there is no test file for `@nivo/network` so added a new one with some tests related to this change.

Adding in user svg:

Adding in a permanent text component below original Node:

EDIT: This change has also added support for custom node components for NodeCanvas.

### CODE REVIEW

File: packages/network/src/Node.js [hunk 1]
```diff
@@ -48,4 +48,4 @@ Node.propTypes = {
     handleNodeLeave: PropTypes.func.isRequired,
 }
 
-export default memo(Node)
+export default Node
```
Comment (plouc) [inline packages/network/src/Node.js:51]: Why did you remove the memoization?

File: packages/network/src/NetworkCanvas.js [hunk 2]
```diff
@@ -60,6 +61,7 @@ const NetworkCanvas = props => {
 
     const theme = useTheme()
     const getNodeColor = useNodeColor(nodeColor)
+    const getNodeComponent = useNodeComponent(nodeComponent)
```
Comment (plouc) [inline packages/network/src/NetworkCanvas.js:64]: This isn't used currently, you can limit this to the SVG implementation, or add support for canvas, but you'll have to adjust the TS definitions and prop-types, also a custom node component for canvas would be a function receiving a 2D canvas context as the first argument and the _props_ as the second one. The swarmplot package use this approach if you're curious, and given you're willing to implement it of course :)

File: packages/network/tests/Network.test.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    },
+]
+const links = [
+    {
+        source: '0',
+        target: '1',
+    },
+]
+
+describe('Network', () => {
```
Comment (plouc) [inline packages/network/tests/Network.test.tsx:28]: Thank you!

File: website/src/data/components/network/props.js [hunk 4]
```diff
@@ -137,7 +137,7 @@ const props = [
     {
         key: 'nodeBorderWidth',
         group: 'Nodes',
-        type: 'number | (node: Node) => number',
+        type: 'Function',
```
Comment (plouc) [inline website/src/data/components/network/props.js:140]: It can be either a number or a function, a function is always used internally for consistency, but you can use a static number as in the website demo for example.

File: website/src/data/components/network/props.js [hunk 5]
```diff
@@ -150,6 +150,12 @@ const props = [
         defaultValue: NetworkDefaultProps.nodeBorderColor,
         controlType: 'inheritedColor',
     },
+    {
+        key: 'nodeComponent',
+        group: 'Nodes',
+        type: '(node: Node) => svg',
```
Comment (plouc) [inline website/src/data/components/network/props.js:156]: It's not correct if we consider canvas, you can add `flavors: ['svg'],` if you remove support for canvas, or adjust the type to use a union otherwise.

**Review by plouc — CHANGES_REQUESTED** [review]

**Comment by plouc** [comment]
@SheldonNunes, thank you for this, this could be really useful! and sorry for the late review.

File: packages/network/src/hooks.js [hunk 6]
```diff
@@ -106,6 +106,12 @@ export const useNodeColor = color =>
         return () => color
     }, [color])
 
+export const useNodeComponent = nodeComponent =>
```
[the remark this answers is not in the document]
Comment (SheldonNunes) [reply packages/network/src/hooks.js:109]: You are right. I had a misunderstanding of why these hooks were being used. I have removed in favour of passing in the component as you suggested

File: packages/network/src/Node.js (hunk 1 above)
Comment (SheldonNunes) [reply packages/network/src/Node.js]: Not sure... I have put it back and there is no issue 👍

File: packages/network/src/NetworkCanvas.js (hunk 2 above)
Comment (SheldonNunes) [reply packages/network/src/NetworkCanvas.js]: I have updated this PR with this suggestion. Let me know if there is any improvements for the the Canvas implementation 🙏

File: website/src/data/components/network/props.js (hunk 5 above)
Comment (SheldonNunes) [reply website/src/data/components/network/props.js]: Similar to the swarmplot implementation I have added a `nodeComponent` for svg flavor and `renderNode` for canvas flavor.

Comment (plouc) [reply website/src/data/components/network/props.js]: Makes sense!

**Comment by plouc** [comment]
@SheldonNunes, as I am currently converting `@nivo/network` to TypeScript: https://github.com/plouc/nivo/pull/1767, I went ahead and added this feature when converting, I wasn't sure you had the time to address the feedbacks, sorry.

**Comment by plouc** [comment]
It's not yet exposed as a property in the PR, but already supported via `createElement`: https://github.com/plouc/nivo/pull/1767/files#diff-e3d2cb937239fcccb5e5a3df4d21fd3af0ce57b457285f18b3147d9ec35bd733R75, I will reuse what you did for canvas.

This package will now also use `react-spring`, because of https://github.com/plouc/nivo/issues/667, this changes the API a little bit, and impacts the internal implementation a lot.

**Comment by SheldonNunes** [comment]
Ah ok! Shall I close off this PR then?

**Comment by plouc** [comment]
@SheldonNunes, yes, again, sorry for that, but it is useful to add the feature in the other branch, so thank you!

**Comment by SheldonNunes** [comment]
No worries. I had kind of forgotten about it until I got pinged it was reopened and had some extra time :) 

Glad this feature will make it into nivo though! :D

**Comment by plouc** [comment]
@SheldonNunes, also adding support for custom links, could be useful if we want dashed lines, arcs instead of straight lines or event labels.
